### DISCUSSION TITLE
Abstract Operations on Discrete Data

### DISCUSSION THREAD

**Discussion by s-aucoin** [CliMA/Oceananigans.jl#3211] [Computational science!] [answered]
Hi,

I have a question about using abstract operations on discrete data, specifically FieldTimeSeries objects. I understand that you can use these operations (like the spatial derivatives) before running a simulation to calculate and store additional variables like the following:
``` 
u, v, w = model.velocities
         ξ = (∂y(w) - ∂z(v))                                  # (1/s) x-z vorticity
         η = (∂z(u) - ∂x(w))                                  # (1/s) y-z vorticity
         ζ = (∂x(v) - ∂y(u))                                  # (1/s) x-y vorticity
      vort = (; ξ, η, ζ)

simulation.output_writers[:slices] =
    JLD2OutputWriter(model, merge(model.velocities, model.tracers, vort),
                     filename = filepath,
                     schedule = TimeInterval(output_interval),
                     overwrite_existing = true)
```

Due mainly to storage concerns, I only save the basic variables in my output file, so my question is can the same thing (or an analogous operation) be done after the simulation with the FieldTimeSeries data? Something like:
```
time_series = (u = FieldTimeSeries(load_filepath, "u"),            # (m s⁻¹) x component of velocity
               v = FieldTimeSeries(load_filepath, "v"),            # (m s⁻¹) y component of velocity
               w = FieldTimeSeries(load_filepath, "w"))            # (m s⁻¹) z component of velocity

extra_time_series = (ξ = (∂y(time_series.w) - ∂z(time_series.v)),                        # (s⁻¹) x-z vorticity
                     η = (∂z(time_series.u) - ∂x(time_series.w)),                        # (s⁻¹) y-z vorticity
                     ζ = (∂x(time_series.v) - ∂y(time_series.u)))                        # (s⁻¹) x-y vorticity

new_time_series = merge(time_series, extra_time_series)
```

I know I could define a discrete derivative myself separately, but I would like to ensure consistency in the discretization, and ideally would like to include the whole domain (using the boundary conditions). Is this possible?

**Comment by glwagner** [comment, marked as answer]
Yes, except that you cannot perform `AbstractOperations` directly on `FieldTimeSeries`. Instead, you have to extract a `Field` at a particular time from the `FieldTimeSeries` and perform operations on that:

```julia
ut = FieldTimeSeries(filename, "u")
vt = FieldTimeSeries(filename, "v")

for n = 1:length(ut.times)
    un = ut[n]
    vn = vt[n]
    ζn = ∂x(vn) - ∂y(un)
end
```

Also note that you probably want to save your data using `with_halo = true` so that you have halo points (which will allow you to compute derivatives on the boundaries).

**Reply by s-aucoin** [reply]
Thanks! How would I then recombine the `Fields` into a `FieldTimeSeries`?

**Reply by glwagner** [reply]
For that, you can make an empty `FieldTimeSeries`. For vorticity this would look lik

```julia
ζt = FieldTimeSeries{Face, Face, Center}(grid, ut.times)
```

You might be able to fill the interior values with:

```julia
for n = 1:length(ut.times)
    un = ut[n]
    vn = vt[n]
    ζt[n] .= ∂x(vn) - ∂y(un)
end
```

I'm not 100% sure that would work. Let us know.

PS I think the docstring for `FieldTimeSeries` should have some of this info, but if not perhaps we should improve it?

**Reply by s-aucoin** [reply]
I tested this approach, and it worked on a more powerful machine I use, but when I try to do it on my laptop, the Julia kernel consistently dies when trying to run `ζt[n] .= ∂x(vn) - ∂y(un)`. Could this be a memory issue?

Thanks for the help.

**Reply by glwagner** [reply]
What do you mean by "dies"?

**Reply by glwagner** [reply]
You could try something like

```julia
un = XFaceField(grid)
vn = YFaceField(grid)
ζn = Field(∂x(vn) - ∂y(un))

for n = 1:length(ut.times)
    parent(un) .= parent(ut[n])
    parent(vn) .= parent(vt[n])
    compute!(ζn)
    parent(ζt[n]) .= parent(ζn)
end
```

not sure that will work or help...

**Reply by s-aucoin** [reply]
"Dying" is just the terminology that Jupyter uses to say it crashed. This other solution seems to have the same result; it works, but the script crashes when trying to run `compute!(ζn)`. This makes me think that it must be running out of memory.

Thank you for your help, I will try to find a workaround.

**Reply by glwagner** [reply]
A quick search shows that jupyter kernels can quit when they run out of memory, so that could make sense. Oddly, that must have to do with the memory used during compilation (not the memory consumed by the arrays), since otherwise it would not quit during `compute!`. I wonder if it works at the REPL.

**Reply by s-aucoin** [reply]
I tried running the script as a .jl file using the REPL, but it also crashed. It's strange because it doesn't seem like a very memory heavy task, and my computer has a decent amount (16 GB).

**Reply by glwagner** [reply]
How does the REPL crash? Do you mean that it throws an error?

**Reply by s-aucoin** [reply]
It does throw an error, but then immediately closes the REPL. Running it in VSCode does the same thing but also gives the message that it exits with exit code 1.

**Reply by glwagner** [reply]
I've never seen that before! Crazy!

I have seen `Out of memory` errors. However, another thing that can happen is that your CPU starts using "swap memory" (which is on disk and extremely slow). Then things will continue to run but probably would never finish because its so slow.
